pkg/printers/jsonpath_flags.go

```diff
@@ -0,0 +1,121 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package printers
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// JSONPathPrintFlags provides default flags necessary for template printing.
+// Given the following flag values, a printer can be requested that knows
+// how to handle printing based on these values.
+type JSONPathPrintFlags struct {
+	// indicates if it is OK to ignore missing keys for rendering
+	// an output template.
+	AllowMissingKeys *bool
+	TemplateArgument *string
+}
+
+// ToPrinter receives an outputFormat and returns a printer capable of
+// handling --template format printing.
+// Returns false if the specified outputFormat does not match a template format.
+func (f *JSONPathPrintFlags) ToPrinter(outputFormat string) (ResourcePrinter, bool, error) {
+	if (f.TemplateArgument == nil || len(*f.TemplateArgument) == 0) && len(outputFormat) == 0 {
+		return nil, false, fmt.Errorf("missing --template value")
+	}
+
+	templateArg := ""
+
+	// templates are logically optional for specifying a format.
+	// this allows a user to specify a template format value
+	// as --output=jsonpath=
+	templateFormats := map[string]bool{
+		"jsonpath":      true,
+		"jsonpath-file": true,
+	}
+
+	if f.TemplateArgument == nil || len(*f.TemplateArgument) == 0 {
+		for format := range templateFormats {
+			format = format + "="
+			if strings.HasPrefix(outputFormat, format) {
+				templateArg = outputFormat[len(format):]
+				outputFormat = format[:len(format)-1]
+				break
+			}
+		}
+	} else {
+		templateArg = *f.TemplateArgument
+	}
+
+	if _, supportedFormat := templateFormats[outputFormat]; !supportedFormat {
+		return nil, false, nil
+	}
+
+	if len(templateArg) == 0 {
+		return nil, true, fmt.Errorf("template format specified but no template given")
+	}
+
+	if templateArg == "jsonpath-file" {
+		data, err := ioutil.ReadFile(templateArg)
+		if err != nil {
+			return nil, true, fmt.Errorf("error reading --template %s, %v\n", templateArg, err)
+		}
+
+		templateArg = string(data)
+	}
+
+	p, err := NewJSONPathPrinter(templateArg)
+	if err != nil {
+		return nil, true, fmt.Errorf("error parsing jsonpath %s, %v\n", templateArg, err)
+	}
+
+	allowMissingKeys := true
+	if f.AllowMissingKeys != nil {
+		allowMissingKeys = *f.AllowMissingKeys
+	}
+
+	p.AllowMissingKeys(allowMissingKeys)
+	return p, true, nil
+}
+
+// AddFlags receives a *cobra.Command reference and binds
+// flags related to template printing to it
+func (f *JSONPathPrintFlags) AddFlags(c *cobra.Command) {
+	if f.TemplateArgument != nil {
+		c.Flags().StringVar(f.TemplateArgument, "template", *f.TemplateArgument, "Template string or path to template file to use when --output=jsonpath, --output=jsonpath-file.")
```

Comment:
Do we need it to be pointer? We can always check `Changed` value to see if it was set, but for majority of cases this only complicates the code.

Reply:
> Do we need it to be pointer? We can always check Changed value to see if it was set, but for majority of cases this only complicates the code.

Once you compose the values, you can no longer guarantee you were set by cobra and no code after the binding should ever assume it has access to the flagset.